internal/vcs/metadata.go

```diff
@@ -0,0 +1,244 @@
+
+// GetMetadata fetches vcs metadata for the given environment. GetMetadata will attempt to try and find
+// a GetMetadataFunc for the environment using env variables to determine the CI system. If GetMetadata
+// cannot determine the CI system it falls back to getting the metadata from the local git filesystem.
+func GetMetadata() (Metadata, error) {
+	if isTest() {
+		return StubMetadata, nil
+	}
+
+	for e, f := range envToProviderMap {
+		_, ok := os.LookupEnv(e)
+		if ok {
+			return f()
+		}
+	}
+
+	return getLocalGitMetadata()
+}
+
+func isTest() bool {
+	return os.Getenv("INFRACOST_ENV") == "test" || strings.HasSuffix(os.Args[0], ".test")
+}
+
+func getGitlabMetadata() (Metadata, error) {
+	m, err := getLocalGitMetadata()
+	if err != nil {
+		return m, fmt.Errorf("github metdata error, could not fetch initial metadata from local git %w", err)
+	}
+
+	if m.Branch.Name == "HEAD" {
+		m.Branch.Name = os.Getenv("CI_MERGE_REQUEST_SOURCE_BRANCH_NAME")
+	}
+
+	m.Pipeline = &Pipeline{ID: os.Getenv("CI_JOB_ID")}
+	m.PullRequest = &PullRequest{
+		VCSProvider:  "gitlab-ci",
+		Title:        os.Getenv("CI_MERGE_REQUEST_TITLE"),
+		Author:       os.Getenv("CI_COMMIT_AUTHOR"),
+		SourceBranch: os.Getenv("CI_MERGE_REQUEST_SOURCE_BRANCH_NAME"),
+		BaseBranch:   os.Getenv("CI_MERGE_REQUEST_TARGET_BRANCH_NAME"),
+		URL:          os.Getenv("CI_MERGE_REQUEST_PROJECT_URL") + "/-/merge_requests/" + os.Getenv("CI_MERGE_REQUEST_IID"),
+	}
+
+	return m, nil
+}
+
+func getGithubMetadata() (Metadata, error) {
+	event, err := os.ReadFile(os.Getenv("GITHUB_EVENT_PATH"))
+	if err != nil {
+		return Metadata{}, fmt.Errorf("could not read the github event file %w", err)
+	}
+
+	m, err := getLocalGitMetadata()
+	if err != nil {
+		return m, fmt.Errorf("github metdata error, could not fetch initial metadata from local git %w", err)
+	}
+
+	// if the branch name is HEAD this means that we're using a merge commit and need
+	// to fetch the actual commit.
+	if m.Branch.Name == "HEAD" {
+		r, err := git.PlainOpen(".")
+		if err != nil {
+			return Metadata{}, fmt.Errorf("could not open git directory to fetch metadata %w", err)
+		}
+
+		cnf, err := r.Config()
+		if err != nil {
+			return Metadata{}, fmt.Errorf("error opening the .git/config folder in github %w", err)
+		}
+
+		auth := cnf.Raw.Section("http").Subsection("https://github.com/").Options.Get("extraheader")
+		val := strings.TrimSpace(strings.ReplaceAll(auth, "AUTHORIZATION: basic", ""))
+		b, err := base64.URLEncoding.DecodeString(val)
+		if err != nil {
+			return Metadata{}, fmt.Errorf("github basic auth credentials were malformed, could not decode %w", err)
+		}
+
+		pieces := strings.Split(string(b), ":")
+		if len(pieces) != 2 {
+			return Metadata{}, fmt.Errorf("github basic auth credentials were malformed invalid auth components %+v", pieces)
+		}
+
+		headRef := gjson.GetBytes(event, "pull_request.head.ref").String()
+		clonePath := fmt.Sprintf("/tmp/%s", headRef)
```

Comment:
Maybe worth adding the repo name in the path, not just the commit sha? Or at least prefix it with `infracost-`. What do you think?

Reply:
👍

Reply:
done in https://github.com/infracost/infracost/pull/1758/commits/d891b689c8a5d62dc85708d6555ea6ffa1373f35